tensorflow/contrib/igfs/kernels/igfs_messages.h

```diff
@@ -0,0 +1,356 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_CONTRIB_IGFS_KERNELS_IGFS_MESSAGES_H_
+#define TENSORFLOW_CONTRIB_IGFS_KERNELS_IGFS_MESSAGES_H_
+
+#include "tensorflow/contrib/igfs/kernels/igfs_extended_tcp_client.h"
+
+namespace tensorflow {
+
+enum CommandId {
+  HANDSHAKE_ID = 0,
+  EXISTS_ID = 2,
+  INFO_ID = 3,
+  RENAME_ID = 6,
+  DELETE_ID = 7,
+  MKDIR_ID = 8,
+  LIST_PATHS_ID = 9,
+  LIST_FILES_ID = 10,
+  OPEN_READ_ID = 13,
+  OPEN_APPEND_ID = 14,
+  OPEN_CREATE_ID = 15,
+  CLOSE_ID = 16,
+  READ_BLOCK_ID = 17,
+  WRITE_BLOCK_ID = 18,
+};
+
+class IGFSPath {
+ public:
+  Status Read(ExtendedTCPClient *client);
+
+  string path;
+};
+
+class IGFSFile {
+ public:
+  Status Read(ExtendedTCPClient *client);
+
+  int64_t length;
+  int64_t modification_time;
+  uint8_t flags;
+};
+
+class Request {
+ public:
+  Request(int32_t command_id);
+  virtual Status Write(ExtendedTCPClient *client) const;
+
+ protected:
+  const int32_t command_id_;
+};
+
+class Response {
+ public:
+  virtual Status Read(ExtendedTCPClient *client);
+
+  int32_t res_type;
+  int32_t req_id;
+  int32_t length;
+
+ protected:
+  static const int32_t header_size_ = 24;
+  static const int32_t response_header_size_ = 9;
+};
+
+class PathCtrlRequest : public Request {
+ public:
+  PathCtrlRequest(int32_t command_id, const string &user_name,
+                  const string &path, const string &destination_path, bool flag,
+                  bool collocate, const std::map<string, string> &properties);
+  Status Write(ExtendedTCPClient *client) const override;
+
+ protected:
+  Status WritePath(ExtendedTCPClient *client, const string &path) const;
+
+  const string user_name_;
+  const string path_;
+  const string destination_path_;
+  const bool flag_;
+  const bool collocate_;
+  const std::map<string, string> props_;
+};
+
+class StreamCtrlRequest : public Request {
+ public:
+  StreamCtrlRequest(int32_t command_id, int64_t stream_id, int32_t length);
+  Status Write(ExtendedTCPClient *client) const override;
+
+ protected:
+  int64_t stream_id_;
+  int32_t length_;
+};
+
+template <class R>
+class CtrlResponse : public Response {
+ public:
+  CtrlResponse(bool optional) : optional_(optional){};
+  Status Read(ExtendedTCPClient *client) override {
+    TF_RETURN_IF_ERROR(Response::Read(client));
+
+    if (optional_) {
+      TF_RETURN_IF_ERROR(client->ReadBool(&has_content));
+
+      if (!has_content) return Status::OK();
+    }
+
+    res = R();
+    has_content = true;
+    TF_RETURN_IF_ERROR(res.Read(client));
+
+    return Status::OK();
+  }
+
+  R res;
+  bool has_content;
+
+ private:
+  bool optional_;
+};
+
+template <class T>
+class ListResponse {
+ public:
+  Status Read(ExtendedTCPClient *client) {
+    int32_t len;
+    TF_RETURN_IF_ERROR(client->ReadInt(&len));
+
+    entries = std::vector<T>();
```

Comment:
Suggestion: `entries.clear()` is slightly more readable, since the object must already have been constructed.

Reply:
Done.